### PR TITLE
Latest changes to compile with Etoile, but it has runtime error.

### PR DESCRIPTION
**Pull request by huperniketes**
For issue #1 
Added CoreObject framework to EtoilePaint target.
Code changes to compile without error with latest EtoileUI.

It has runtime error: unknown selector on +[ETBezierFreeLayout layout].

### CODE REVIEW

File: ETBezierHandle.m
Comment (qmathe) [inline ETBezierHandle.m]: Should be [self valueForVariableStorageKey: kETManipulatedObjectProperty]

_COObject_ should be _self_ everywhere you replaced GET/SET_PROPERTY.

File: Sketch/SketchController.h
Comment (qmathe) [inline Sketch/SketchController.h]: A controller should be a subclass of ETController only if it is assigned to ETLayoutItem.controller somewhere.

Here SketchController is an app controller that builds the UI and puts on screen, so it should remain a NSObject subclass.

File: Sketch/SketchController.m
Comment (qmathe) [inline Sketch/SketchController.m]: This should be removed when changing the superclass back to NSObject.

File: Sketch/SketchController.m
Comment (qmathe) [inline Sketch/SketchController.m]: [self objectGraphContext] should be [itemFactory objectGraphContext]

ETLayoutItemFactory.objectGraphContext is the object graph of the UI that is currently built with the item factory.

The object graph context represents a document object graph. Here we can consider the UI we built makes up a single UI document (similar to a Nib file), where the root node is mainItem.

File: Sketch/SketchController.m
Comment (qmathe) [inline Sketch/SketchController.m]: [self objectGraphContext] should be [itemFactory objectGraphContext]

File: Sketch/SketchController.m
Comment (qmathe) [inline Sketch/SketchController.m]: [self objectGraphContext] should be [itemFactory objectGraphContext]

**Review by qmathe — CHANGES_REQUESTED** [review]
Thanks for working on this! I have made some comments. If something is unclear or you want some explanations, let me know.

To better understand how CoreObject works and what a COObjectGraphContext represents, you should take a look at:
- http://coreobject.org/technotes/ (you can skip Undo, Diff/Merge and Collaborative Editing sections)
- http://etoileos.com/dev/api/CoreObject/COEditingContext.html#COEditingContext

File: Sketch/SketchController.m
Comment (huperniketes) [reply Sketch/SketchController.m]: If there's an object graph specific to the document's UI, can the item factory be used anywhere the object graph needs to be accessed? If so, I should be able to access the graph within ETBezierHandler's initializer like so `[[ETLayoutItemFactory factory]  objectGraphContext]`, correct?

Comment (qmathe) [reply Sketch/SketchController.m]: No, for both questions. The item factory is only for building the UI (the item tree), then in the implementation code of any ETUIObject subclass (e.g. ETLayoutItem, ETLayout, ETBezierPathHandle), the UI object graph should be accessed with self.objectGraphContext.  Similarly if you want to know the object graph of a bezier path handle, you should access it with handle.objectGraphContext and not with -[ETLayoutItemFactory factory].objectGraphContext.

If the UI happens to be cloned or a bezier path handle is copied into another object graph, [ETLayoutItemFactory factory].objectGraphContext won't work anymore, since the copies  will belong to another object graph.

-[ETLayoutItemFactory factory] is just a convenience method to build a non-persistent UI. For persistent UIs, the factory must be created with +factoryWithObjectGraphContext: with  an object graph context that belongs to a COPersistentRoot. Each persistent root represents a distinct document in a CoreObject store.

Comment (huperniketes) [reply Sketch/SketchController.m]: So in updating -[ETBezierHandle initWithActionHandler:manipulatedObject:partcode:] to invoke the super's -initWithActionHandler:manipulatedObject:objectGraphContext:, where do I get the correct graph context? Does that parameter need to be added to ETBezierHandle's initializer too?

**Comment by huperniketes** [comment]
Does this look right now?

File: Sketch/SketchController.m
Comment (qmathe) [reply Sketch/SketchController.m]: Yes, that's right. For COObject subclasses, the designated initializer should always take an object graph context as its last argument. So classes like ETBezierHandle and ETBezierHandleGroup should have their designated initializer changed.

**Comment by qmathe** [comment]
Everything looks good now :-) If you want to correct ETBezierHandle and ETBezierHandleGroup in this pull request, I can wait before merging it. Let me know what you prefer.

**Comment by huperniketes** [comment]
I fixed the initializers for ETBezierHandle and ETBezierHandleGroup as you asked, and I think the accessor -[ETBezierPointActionHandler sharedInstance] needs to be updated as well, correct?

**Comment by qmathe** [comment]
Yes, all +sharedInstance methods in EtoilePaint should be removed when they belong to ETUIObject subclasses.

`+[ETUIObject sharedInstanceForObjectGraphContext:]` supercedes it. It is inherited by all ETUIObject subclasses and existing +sharedInstance callers should be updated to use it.

File: ETBezierHandle.h
Comment (qmathe) [inline ETBezierHandle.h]: objectGraphContext: must always appear as the last argument in an initializer. It's an initializer convention defined by CoreObject for all COObject subclasses.

File: ETBezierHandle.m
Comment (qmathe) [inline ETBezierHandle.m]: Would be better to use space to align method keywords, otherwise they won't appear correctly aligned with a custom tab size like in Github.

I know it's annoying since Xcode always inserts tabs instead :(

In the long run, I'll probably convert EtoilePaint to use spaces instead of tabs, for a better Xcode compatibility.

File: ETBezierHandle.m
Comment (qmathe) [inline ETBezierHandle.m]: Indenting with spaces would be better here

File: ETBezierHandle.m
Comment (qmathe) [inline ETBezierHandle.m]: Same indentation issue here. Indenting ETBezierHandler calls with spaces instead of tabs would be better.

File: ETBezierEditingFreeLayout.m
Comment (qmathe) [inline ETBezierEditingFreeLayout.m]: I just tried to compile your code, but this import is breaking the compilation (NSView+EtoileUI.h is private). If you remove it, the code is compiles fine.

File: ETBezierHandle.h
Comment (qmathe) [inline ETBezierHandle.h:44]: ETBezierEditingFreeLayout implementation needs to be updated to use -initWithManipulatedObject:objectGraphContext: to eliminate the following warning:

Etoile/Frameworks/EtoilePaint/ETBezierEditingFreeLayout.m:26:72: Instance method '-initWithManipulatedObject:' not found (return type defaults to 'id')

File: ETBezierHandle.m
Comment (qmathe) [inline ETBezierHandle.m]: In this case block, the casts to COObjectGraphContext look useless

File: ETBezierHandle.m
Comment (qmathe) [inline ETBezierHandle.m]: To reduce the length of these lines, I'd suggest to assign [ETBezierPointActionHandler sharedInstanceForObjectGraphContext: aContext] to a local variable and passing it as an argument

**Review by qmathe — CHANGES_REQUESTED** [review]
Here are some more comments based on the warnings/errors I got when I compiled your changes.

File: ETBezierEditingFreeLayout.m
Comment (huperniketes) [reply ETBezierEditingFreeLayout.m]: Yes. I changed it to use the same facility that was moved to EtoileUI, but now that the framework provides another mechanism we no longer need this at all.

File: ETBezierHandle.m
Comment (huperniketes) [reply ETBezierHandle.m]: I also wrapped the first case in a loop to eliminate the unnecessary duplication, possibly resolving the FIXME.

**Comment by qmathe** [comment]
Thanks a lot for all these changes! Everything was looking good, so I merged your pull request.

**Pull request merged**
